test/distributed/elastic/rendezvous/utils_test.py

```diff
@@ -215,3 +220,103 @@ def test_matches_machine_hostname_returns_false_if_hostname_does_not_match(self)
         for host in hosts:
             with self.subTest(host=host):
                 self.assertFalse(_matches_machine_hostname(host))
+
+
+class PeriodicTimerTest(TestCase):
+    def test_start_can_be_called_only_once(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        with self.assertRaisesRegex(RuntimeError, r"^The timer has already started.$"):
+            timer.start()
+
+        timer.cancel()
+
+    def test_cancel_can_be_called_multiple_times(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        timer.cancel()
+        timer.cancel()
+
+    def test_cancel_stops_background_thread(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        self.assertTrue(any(t.name == "PeriodicTimer" for t in threading.enumerate()))
+
+        timer.cancel()
+
+        self.assertTrue(all(t.name != "PeriodicTimer" for t in threading.enumerate()))
+
+    def test_delete_stops_background_thread(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        self.assertTrue(any(t.name == "PeriodicTimer" for t in threading.enumerate()))
+
+        del timer
+
+        self.assertTrue(all(t.name != "PeriodicTimer" for t in threading.enumerate()))
+
+    def test_timer_calls_background_thread_at_regular_intervals(self):
+        begin_time = time.monotonic()
+
+        # Call our function every 200ms.
+        call_interval = 0.2
+
+        # Keep the log of intervals between each consecutive call.
+        actual_call_intervals: List[float] = []
+
+        # Keep the number of times the function was called.
+        call_count = 0
+
+        # In order to prevent a flaky test instead of asserting that the
+        # function was called an exact number of times we use a lower bound
+        # that is guaranteed to be true for a correct implementation.
+        min_required_call_count = 4
+
+        timer_stop_event = threading.Event()
+
+        def log_call(self):
+            nonlocal begin_time, call_count
+
+            actual_call_intervals.append(time.monotonic() - begin_time)
+
+            call_count += 1
+            if call_count == min_required_call_count:
+                timer_stop_event.set()
+
+            begin_time = time.monotonic()
+
+        timer = _PeriodicTimer(timedelta(seconds=call_interval), log_call, self)
+
+        timer.start()
+
+        # Although this is theoretically non-deterministic, if our timer, which
+        # has a 200ms call interval, does not get called 4 times in 60 seconds,
+        # there is very likely something else going on.
+        timer_stop_event.wait(60)
```

Comment:
nit: Would there be any way to make this test run faster than 1 minute? For example, divide 200ms by a factor 10 and wait for 6s should be equivalent right? Context is that we are actively trying to reduce the amount of time distributed tests take as per https://github.com/pytorch/pytorch/issues/55583.

Reply:
In normal circumstances this test won't take any longer than ~800ms to complete. You can think of that 60 seconds as a timeout value. Once the `call_count` reaches `min_required_call_count` at line 291, we trigger the event (which should happen at or shortly after the 800ms mark). If you like, I can decrease it from 60 seconds to a lower value (10 seconds maybe?) to make sure that we don't hang too long in case there is something obviously wrong with the test environment. I don't want to make the timeout too short though. The closer it gets to 800ms, the flakier it becomes.

Reply:
Ah ok, I missed that this is an event that's unblocked when the call count is reached.